src/main/java/io/quarkus/gizmo/Type.java

```diff
@@ -0,0 +1,488 @@
+        public static final PrimitiveType BOOLEAN = new PrimitiveType(Primitive.BOOLEAN);
+        public static final PrimitiveType BYTE = new PrimitiveType(Primitive.BYTE);
+        public static final PrimitiveType SHORT = new PrimitiveType(Primitive.SHORT);
+        public static final PrimitiveType INT = new PrimitiveType(Primitive.INT);
+        public static final PrimitiveType LONG = new PrimitiveType(Primitive.LONG);
+        public static final PrimitiveType FLOAT = new PrimitiveType(Primitive.FLOAT);
+        public static final PrimitiveType DOUBLE = new PrimitiveType(Primitive.DOUBLE);
+        public static final PrimitiveType CHAR = new PrimitiveType(Primitive.CHAR);
+
+        private final Primitive primitive;
+
+        PrimitiveType(Primitive primitive) {
+            this.primitive = Objects.requireNonNull(primitive);
+        }
+
+        @Override
+        void appendToSignature(StringBuilder signature) {
+            switch (primitive) {
+                case BOOLEAN:
+                    signature.append("Z");
+                    break;
+                case BYTE:
+                    signature.append("B");
+                    break;
+                case SHORT:
+                    signature.append("S");
+                case INT:
+                    signature.append("I");
+                    break;
+                case LONG:
+                    signature.append("J");
+                    break;
+                case FLOAT:
+                    signature.append("F");
+                    break;
+                case DOUBLE:
+                    signature.append("D");
+                    break;
+                case CHAR:
+                    signature.append("C");
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown primitive type: " + primitive.toString());
+            }
+        }
+
+        @Override
+        boolean isPrimitive() {
+            return true;
+        }
+
+        @Override
+        PrimitiveType asPrimitive() {
+            return this;
+        }
+    }
+
+    public static final class ClassType extends Type {
+        public static final ClassType OBJECT = new ClassType("java/lang/Object", null);
+
+        private final String name; // always slash-delimited
+        private final Type owner;
+
+        ClassType(String name, Type owner) {
+            this.name = Objects.requireNonNull(name);
+            this.owner = owner;
+        }
+
+        /**
+         * Allows building a signature like {@code Lcom/example/Outer.Inner;}. This is usually
+         * unnecessary, because {@code Lcom/example/Outer$Inner} is also a valid signature,
+         * but it's occasionally useful to build a signature of more complex inner types.
+         *
+         * @param simpleName simple name of the member class nested in this class
+         * @return the inner class
+         */
+        public ClassType innerClass(String simpleName) {
```

Comment:
@Ladicek BTW `innerClass`  is IMO inaccurate because only _non-static nested classes_ are called _inner classes_ in Java... so I think it should be `nestedClass()`.

Reply:
For some reason, I got pretty convinced this is only useful for inner classes and not for static nested classes. Let me look again.

Reply:
Hm, let's keep the `inner` prefix for now...